pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,744 @@
+        try:
+            int_g_coeff_map = int_g_cc(expr)
+        except (RuntimeError, AssertionError):
+            # Don't touch this expression, because it's not linear.
+            # FIXME: if there's ever any use case, then we can extract
+            # some IntGs from them.
+            result[i] += expr
+            continue
+        int_gs_by_group = {}
+        for int_g, coeff in int_g_coeff_map.items():
+            if int_g == 1:
+                # coeff map may have some constant terms, add them to
+                result[i] += coeff
+                continue
+
+            # convert DirectionalSourceDerivative to AxisSourceDerivative
+            # as kernel arguments need to be the same for merging
+            if source_dependent_variables is not None:
+                int_g = convert_directional_source_to_axis_source(int_g,
+                    source_dependent_variables)
+            # convert TargetDerivative to source before checking the group
+            # as the target kernel has to be the same for merging
+            int_g = convert_target_deriv_to_source(int_g)
+            # move the coefficient inside
+            int_g = int_g.copy(densities=[density*coeff for density in
+                    int_g.densities])
+
+            source_group = get_int_g_source_group(int_g, differing_kernel_arguments)
+            target_group = get_int_g_target_group(int_g)
+            group = (source_group, target_group)
+
+            all_source_groups[source_group] = 1
+
+            if group not in int_gs_by_group:
+                new_int_g = int_g
+            else:
+                prev_int_g = int_gs_by_group[group]
+                # Let's merge IntGs with the same group
+                new_int_g = merge_two_int_gs(int_g, prev_int_g)
+            int_gs_by_group[group] = new_int_g
+
+        # Do some simplifications after merging. Not stricty necessary
+        for group, int_g in int_gs_by_group.items():
+            # replace an IntG with d axis source derivatives to an IntG
+            # with one directional source derivative
+            # TODO: reenable this later
+            # result_int_g = convert_axis_source_to_directional_source(int_g)
+            # simplify the densities as they may become large due to pymbolic
+            # not doing automatic simplifications unlike sympy/symengine
+            result_int_g = int_g.copy(
+                    densities=simplify_densities(int_g.densities))
+            int_gs_by_group[group] = result_int_g
+
+        int_gs_by_group_for_index.append(int_gs_by_group)
+
+    # If source_dependent_variables (sdv) is not given return early.
+    # Check for (sdv is None) instead of just (sdv) because
+    # it can be an empty list.
+    if source_dependent_variables is None:
+        for iexpr, int_gs_by_group in enumerate(int_gs_by_group_for_index):
+            for int_g in int_gs_by_group.values():
+                result[iexpr] += int_g
+        return result
+
```

Comment:
```suggestion
    # If source_dependent_variables (sdv) is not given return early.
    # Check for (sdv is None) instead of just (sdv) because
    # it can be an empty list.
    if source_dependent_variables is None:
        for iexpr, int_gs_by_group in enumerate(int_gs_by_group_for_index):
            for int_g in int_gs_by_group.values():
                result[iexpr] += int_g
        return result

```